src/corelibs/U2Formats/src/FastaFormat.cpp

```diff
@@ -298,192 +286,173 @@ static void load(IOAdapter *io, const U2DbiRef &dbiRef, const QVariantMap &fs, Q
-        CHECK_EXT(buff[0] == FASTA_HEADER_START_SYMBOL, os.setError(FastaFormat::tr("First line is not a FASTA header")), NULL);
+        QString buf(DocumentFormat::READ_BUFF_SIZE + 1, '\0');
+        skipLeadingWhitesAndComments(reader, os);
+        CHECK_OP(os, nullptr);
 
+        QString headerLine = readHeaderLine(reader, os);
+        CHECK_OP(os, nullptr);
         l.tryAcquire(headerLine.capacity());
-        CHECK_OP(os, NULL);
+        CHECK_OP(os, nullptr);
 
-        //read sequence
+        // Read sequence.
         QByteArray sequence;
-        int predictedSize = 1000;
-        sequence.reserve(predictedSize);
-
-        do {
-            len = io->readUntil(buff, READ_BUFF_SIZE, fastaHeaderStart, IOAdapter::Term_Exclude);
-            CHECK_EXT(!io->hasError(), os.setError(io->errorString()), NULL);
-            CHECK_BREAK(len > 0);
-
-            len = TextUtils::remove(buff, len, TextUtils::WHITES);
-            buff[len] = 0;
-
-            l.tryAcquire(len);
-            CHECK_OP(os, NULL);
-
-            sequence.append(buff);
-        } while (!os.isCoR());
+        while (!os.isCoR() && !reader.atEnd()) {
+            reader.readLine(os, buf, DocumentFormat::READ_BUFF_SIZE);
+            CHECK_OP(os, nullptr);
+            if (isCommentLine(buf)) {
+                continue;    // Skip comment line.
+            }
+            if (isHeaderLine(buf)) {
+                reader.undo(os);
+                CHECK_OP(os, nullptr);
+                break;
+            }
+            // DNASequence in UGENE can contain only 1-byte symbols.
+            // Forcing latin1 conversion here: the local multi-byte symbols will be lost.
+            QByteArray dnaChars = buf.toLatin1();
+            char *dnaCharsBuff = dnaChars.data();
+            int sequenceChunkLength = TextUtils::remove(dnaCharsBuff, dnaChars.length(), TextUtils::WHITES);
+            if (sequenceChunkLength > 0) {
+                l.tryAcquire(sequenceChunkLength);
+                CHECK_OP(os, nullptr);
+                sequence.append(dnaCharsBuff, sequenceChunkLength);
+                CHECK_OP_BREAK(os);
+            }
+            os.setProgress(reader.getProgress());
+        }
         sequence.squeeze();
 
-        DNASequence *seq = new DNASequence(headerLine, sequence);
-        seq->alphabet = AppContext::getDNAAlphabetRegistry()->findById(BaseDNAAlphabetIds::NUCL_DNA_EXTENDED());
-        assert(seq->alphabet != NULL);
-
-        if (!seq->alphabet->isCaseSensitive()) {
-            TextUtils::translate(TextUtils::UPPER_CASE_MAP, const_cast<char *>(seq->seq.constData()), seq->seq.length());
+        QString sequenceName = headerLine.mid(1);
+        auto dnaSequence = new DNASequence(sequenceName, sequence);
+        dnaSequence->alphabet = AppContext::getDNAAlphabetRegistry()->findById(BaseDNAAlphabetIds::NUCL_DNA_EXTENDED());
+        SAFE_POINT(dnaSequence->alphabet != nullptr, "Sequence alphabet is null!", nullptr);
```

Comment:
If you fail here, `dnaSequence` memory leaks. Use `QSharedProinter`

Reply:
Good point. I reworked it to move the check before new()